libraries/AP_Scripting/applets/VTOL-quicktune.lua

```diff
@@ -0,0 +1,387 @@
+
+-- work out vehicle type
+if param:get("Q_A_RAT_RLL_SMAX") then
+   is_quadplane = true
+   atc_prefix = "Q_A"
+   gcs:send_text(0, "Quicktune for quadplane loaded")
+elseif param:get("ATC_RAT_RLL_SMAX") then
+   is_quadplane = false
+   gcs:send_text(0, "Quicktune for multicopter loaded")
+else
+   gcs:send_text(0, "Quicktune unknown vehicle")
+   return
+end
+
+-- get time in sections since boot
+function get_time()
+   return millis():tofloat() * 0.001
+end
+
+local axis_names = { "RLL", "PIT", "YAW" }
+local param_suffixes = { "FF", "P", "I", "D", "SMAX", "FLTT", "FLTD", "FLTE" }
+local stages = { "D", "P" }
+local stage = stages[1]
+local last_stage_change = get_time()
+local last_gain_report = get_time()
+local slew_parm = nil
+local slew_target = 0
+local slew_delta = 0
+
+local axes_done = {}
+
+
+-- create params dictionary indexed by name, such as "RLL_P"
+local params = {}
+local param_saved = {}
+local param_changed = {}
+local need_restore = false
+
+for i, axis in ipairs(axis_names) do
+   for j, suffix in ipairs(param_suffixes) do
+      local pname = axis .. "_" .. suffix
+      params[pname] = bind_param(atc_prefix .. "_" .. "RAT_" .. pname)
+      param_changed[pname] = false
+   end
+end
+
+-- reset to start
+function reset_axes_done()
+   for i, axis in ipairs(axis_names) do
+      axes_done[axis] = false
+   end
+   stage = stages[1]
+end
+
+-- get all current param values into param_saved dictionary
+function get_all_params()
+   for pname in pairs(params) do
+      param_saved[pname] = params[pname]:get()
+   end
+end
+
+-- restore all param values from param_saved dictionary
+function restore_all_params()
+   for pname in pairs(params) do
+      local changed = param_changed[pname] and 1 or 0
+      if param_changed[pname] then
+         params[pname]:set(param_saved[pname])
+         param_changed[pname] = false
+      end
+   end
+end
+
+-- save all param values to storage
+function save_all_params()
+   for pname in pairs(params) do
+      if param_changed[pname] then
+         params[pname]:set_and_save(params[pname]:get())
+         param_saved[pname] = params[pname]:get()
+         param_changed[pname] = false
+      end
+   end
+end
+
+-- setup a default SMAX if zero
+function setup_SMAX()
+   for i, axis in ipairs(axis_names) do
+      local smax = axis .. "_SMAX"
+      if params[smax]:get() <= 0 then
+         adjust_gain(smax, DEFAULT_SMAX)
+      end
+   end
+end
+
+-- setup filter frequencies
+function setup_filters()
+   for i, axis in ipairs(axis_names) do
+      local fltd = axis .. "_FLTD"
+      local fltt = axis .. "_FLTT"
+      local flte = axis .. "_FLTE"
+      adjust_gain(fltt, INS_GYRO_FILTER:get() * FLTT_MUL)
+      adjust_gain(fltd, INS_GYRO_FILTER:get() * FLTD_MUL)
+      if axis == "YAW" then
+         local FLTE = params[flte]
+         if FLTE:get() <= 0.0 or FLTE:get() > YAW_FLTE_MAX then
+            adjust_gain(flte, YAW_FLTE_MAX)
+         end
+      end
+   end
+end
```

Comment:
I don't think you should be changing filters. Users should go and set them with the mission planner quick setup wizard. 

We do have a binding for `configured()` so you could only do this if the params have never been set. So you won't throw away someones setup.

Reply:
This was done on the advice of Leonard. We want this to work with all GCS, not just MissionPlanner

Reply:
Its just a scale factor on gyro filter. In which case we should just set the defaults in the Quadplane code. If the assumption is that the user has changed the gyro filter then its likely they have also changed the PID filters because its all documented on the same page of the wiki. The mission planner wizard is just a lookup on from the graphs on the wiki, again it sets the gyro filter and PID filters. 

Filters are a key part in getting a good tune, since this script does not (yet) tune filter values all it is doing is throwing way any tuning the user might have done. 

As I say, all you need to do is check if any of the filter values are already configured, if they are I don't think you should change them.